File: V1/newtonfractal.py
import numpy as np
import matplotlib.pyplot as plt
import math
from tqdm import tqdm

a = complex(1, 0)
roots = [complex(1, 0), complex(-1/2, math.sqrt(3)/2),
         complex(-1/2, -math.sqrt(3)/2)]
INF = 1e9
DX = 1e-9

# The polynomial P(z) in question
def p(z: complex):  # of the form a(z-r1)(z-r2)...
    val = a
    for i in range(len(roots)):
        val *= (z - roots[i])
    return val

# Derivative of P(z)
def pPrime(z: complex):
    dz = complex(DX, DX)
    return (p(z + dz) - p(z)) / dz

# Return a Newton Approximation after a specified number of iterations
def newtonApproximation(a0: complex, numIterations):
    if (p(a0) == 0):
        return a0
    ai = a0
    for i in range(1, numIterations + 1):
        if (pPrime(ai) == 0):
            return ai
        ai = ai - p(ai) / pPrime(ai)
    return ai

# Finds index of closest root to any complex number z
def closestRootTo(z: complex):
    best = INF
    rootNum = 0
    for i in range(len(roots)):
        root = roots[i]
        dist = (z.real - root.real)**2 + (z.imag - root.imag)**2
        if dist < best:
            best = dist
            rootNum = i
    return rootNum

# Main loop to calculate set sized image array 
def calculateImgPlot(radius, pixels, numIterations):
    x = y = np.linspace(-radius, radius, pixels)
    image = []

    for b in tqdm(y):
        line = []
        for a in x:
            z = complex(a, b)
            line.append(closestRootTo(newtonApproximation(z, numIterations)))

        image.append(line)

    return image


def main():
    radius = int(input("Radius: "))
    pixels = int(input("Pixels: "))
    numIterations = int(input("Iterations: "))
    print("")
    
    image = calculateImgPlot(
        radius=radius, pixels=pixels, numIterations=numIterations)
    extent = [-radius, radius, -radius, radius]
    plt.imshow(image, extent=extent)

    rootX = [i.real for i in roots]
    rootY = [i.imag for i in roots]
    plt.scatter(rootX, rootY, c='#ff0000', marker='.')

    plt.title(f"Iterations: {numIterations}")
    plt.show()


if __name__ == "__main__":
    main()

File: V2/newtonfractal.cpp
#include <complex>
#include <iostream>
#include <fstream>
#include <list>
using namespace std;
using namespace std::complex_literals;

typedef complex<double> cpx;
double INF = 1e9;
double dx = 1e-9;

cpx a = 1.0 + 0.0i;
list<cpx> roots = {1.0 + 0.0i, 0.5 + sqrt(3) / 2, 0.5 - sqrt(3) / 2};

// The main polynomial P(z)
cpx p(cpx z) {
    cpx val = a;
    for (cpx root : roots) {
        val *= (z - root);
    }
    return val;
}

// Derivative of P(z)
cpx pPrime(cpx z) {
    return (p(z + dx) - p(z)) / dx;
}

// Returns a Newton Approximation for the roots after a specified number of iterations
cpx newtonApprox(cpx a0, int numIterations) {
    if (p(a0) == cpx(0, 0)) return a0;

    cpx ai = a0;
    for (int i = 1; i <= numIterations; i++) {
        if (pPrime(ai) == cpx(0, 0)) return ai;
        ai = ai - p(ai) / pPrime(ai);
    }

    return ai;
}

// Finds index of closest root to any complex number z
int closestRoot(cpx z) {
    int index = 0, bestRoot = 0;
    double best = INF;
    for (cpx root : roots) {
        double dist = sqrt(pow(z.real() - root.real(), 2) + pow(z.imag() - root.imag(), 2));
        if (dist < best) {
            best = dist;
            bestRoot = index;
        }
        index++;
    }

    return bestRoot;
}

// Main loop to calculate set sized image array
void calculateImgPlot(int radius, int pixels, int numIterations) {
    double interval = 2 * radius / pixels;
    list<list<int>> image;
    list<double> x, y;
    list<int> line;

    for (double i = 1; i <= pixels; i++) {
        x.push_back(interval * i);
        y.push_back(interval * i);
    }

    for (double b : y) {
        for (double a : x) {
            cpx z(a, b);
            line.push_back(closestRoot(newtonApprox(z, numIterations)));
        }
        image.push_back(line);
        line.clear();
    }

    ofstream dataFile;
    dataFile.open("data.csv")
}

int main() {
    cout << closestRoot(newtonApprox(cpx(1, 2), 5));
}